packages/core/src/routes/interaction/utils/single-sign-on.ts

```diff
@@ -0,0 +1,332 @@
+
+      // Will throw ConnectorError if failed to fetch the provider's config
+      return await connectorInstance.getAuthorizationUrl(
+        payload,
+        async (connectorSession: ConnectorSession) =>
+          assignConnectorSessionResult(ctx, provider, connectorSession)
+      );
+    }
+
+    // TODO: Implement SAML SSO providers logic
+    throw new Error('Not implemented');
+  } catch (error: unknown) {
+    // Catch ConnectorError and re-throw as 500 RequestError
+    if (error instanceof ConnectorError) {
+      throw new RequestError({ code: `connector.${error.code}`, status: 500 }, error.data);
+    }
+
+    throw error;
+  }
+};
+
+type SsoAuthenticationResult = {
+  issuer: string;
+  userInfo: SocialUserInfo;
+};
+
+// Get the user authentication result from the SSO provider
+export const getSsoAuthentication = async (
+  ctx: WithLogContext,
+  { provider }: TenantContext,
+  connectorData: SupportedSsoConnector,
+  data?: unknown
+): Promise<SsoAuthenticationResult> => {
+  const { createLog } = ctx;
+  const { id: connectorId, providerName } = connectorData;
+
+  const log = createLog(`Interaction.SignIn.Identifier.SingleSignOn.Submit`);
+  log.append({ connectorId, data });
+
+  try {
+    // Will throw ConnectorError if the config is invalid
+    const connectorInstance = new ssoConnectorFactories[providerName].constructor(connectorData);
+
+    const issuer = connectorInstance.getIssuer();
+    const userInfo = await connectorInstance.getUserInfo(data, async () =>
+      getConnectorSessionResult(ctx, provider)
+    );
+
+    const result = {
+      issuer,
+      userInfo,
+    };
+
+    log.append({ issuer, userInfo });
+
+    return result;
+
+    // TODO: Implement SAML SSO providers logic
+  } catch (error: unknown) {
+    // Catch ConnectorError and re-throw as 500 RequestError
+    if (error instanceof ConnectorError) {
+      throw new RequestError({ code: `connector.${error.code}`, status: 500 }, error.data);
+    }
+
+    throw error;
+  }
+};
+
+// Handle the SSO authentication result and return the user id
+export const handleSsoAuthentication = async (
+  ctx: WithLogContext & WithInteractionHooksContext,
+  tenant: TenantContext,
+  connectorData: SupportedSsoConnector,
+  ssoAuthentication: SsoAuthenticationResult
+): Promise<string> => {
+  const { createLog } = ctx;
+  const { provider, queries } = tenant;
+  const { userSsoIdentities: userSsoIdentitiesQueries, users: usersQueries } = queries;
+  const { issuer, userInfo } = ssoAuthentication;
+
+  // Get logto user info
+  const userSsoIdentity = await userSsoIdentitiesQueries.findUserSsoIdentityBySsoIdentityId(
+    issuer,
+    userInfo.id
+  );
+
+  // SignIn
+  if (userSsoIdentity) {
+    return signInWithSsoAuthentication(ctx, queries, {
+      connectorData,
+      userSsoIdentity,
+      ssoAuthentication,
+    });
+  }
+
+  const user = userInfo.email && (await usersQueries.findUserByEmail(userInfo.email));
```

Comment:
nit: just want to double check: our SSO implementation requires and only uses `email` as the identifier?

Reply:
First of all, same as social, we uses SSO identities' userId as the SSOT to identify a user in Logto. 
In the product perspective, the Single Sign On flow was built based on the email identifier only. Only when user enters a email end with SSO registered domain will we trigger the Single Sign On flow. 
Specifically in this code block case, yes,   we will and will only identify the SSO identities' email and compare with our existing account, link them if a match is found.